File: src/content/docs/topics/cs/coding-problems/graphs/127-word-ladder.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func ladderLength(beginWord string, endWord string, wordList []string) int {
	// TODO: implement
	return 0
}

func runTests() {
	assert(ladderLength("hit", "cog", []string{"hot", "dot", "dog", "lot", "log", "cog"}) == 5)
	assert(ladderLength("hit", "cot", []string{"hot", "dot", "dog", "lot", "log", "cog"}) == 0)
	assert(ladderLength("hot", "dot", []string{"dot", "lot"}) == 2)
	assert(ladderLength("hit", "cog", []string{"hot", "dot", "dog", "lot", "log"}) == 0)
	assert(ladderLength("a", "c", []string{"a", "b", "c"}) == 2)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/graphs/127-word-ladder.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function ladderLength(beginWord: string, endWord: string, wordList: string[]): number {
    // TODO: implement
    return 0;
}

function _runTests(): void {
    assert(ladderLength('hit', 'cog', ['hot','dot','dog','lot','log','cog']) === 5);
    assert(ladderLength('hit', 'cot', ['hot','dot','dog','lot','log','cog']) === 0);
    assert(ladderLength('hot', 'dot', ['dot','lot']) === 2);
    assert(ladderLength('hit', 'cog', ['hot','dot','dog','lot','log']) === 0);
    assert(ladderLength('a', 'c', ['a','b','c']) === 2);
    // perf
    const bigList = Array.from({ length: 10_000 }, (_, i) => String.fromCharCode(97 + (i % 26)).repeat(3) + String.fromCharCode(97 + Math.floor(i / 26) % 26));
    const t0 = performance.now();
    ladderLength('aaaa', 'zzzz', bigList);
    console.log(`perf word-ladder 10000-word list: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
